src/command.ts

```diff
@@ -1,97 +1,95 @@
-  CommandMiddleware,
-  Composer,
+  BotCommandScopeAllChatAdministrators,
+  BotCommandScopeAllGroupChats,
+  BotCommandScopeAllPrivateChats,
+  ChatTypeMiddleware,
   Context,
-  match,
   Middleware,
-  P,
+  MiddlewareObj,
 } from "./deps.deno.ts";
+import { Composer, match, P } from "./deps.deno.ts";
 
 export type MaybeArray<T> = T | T[];
+type BotCommandGroupsScope = BotCommandScopeAllGroupChats | BotCommandScopeAllChatAdministrators;
 
-export class Command<C extends Context = Context> extends Composer<C> {
-  #scopes: BotCommandScope[] = [];
-  #languages: Map<string, { name: string; description: string }> = new Map();
+const isAdmin = (ctx: Context) =>
+  ctx.getAuthor().then((author) => ["administrator", "creator"].includes(author.status));
 
-  constructor(name: string, description: string, ...middleware: Array<Middleware<C>>) {
-    super();
+export class Command<C extends Context = Context> implements MiddlewareObj<C> {
+  private _scopes: BotCommandScope[] = [];
+  private _languages: Map<string, { name: string; description: string }> = new Map();
+  private _composer: Composer<C> = new Composer<C>();
 
-    if (!name.match(/[a-z0-9_]{1,32}/)) {
-      throw new Error(`${name} is not a valid command name`);
-    }
-
-    this.#languages.set("default", { name, description });
+  constructor(name: string, description: string) {
+    this._languages.set("default", { name, description });
+  }
 
-    this.command(name, ...middleware);
+  get scopes() {
+    return this._scopes;
   }
 
   get languages() {
-    return this.#languages;
+    return this._languages;
   }
-  get scopes() {
-    return this.#scopes;
+
+  get names() {
+    return Array.from(this._languages.values()).map(({ name }) => name);
   }
 
   get name() {
-    return this.#languages.get("default")!.name;
+    return this._languages.get("default")!.name;
   }
 
   get description() {
-    return this.#languages.get("default")!.description;
+    return this._languages.get("default")!.description;
   }
 
-  public addToScope(type: "default" | "all_private_chats" | "all_group_chats" | "all_chat_administrators"): this;
-  public addToScope(type: "chat", chatId: string | number): this;
-  public addToScope(type: "chat_member", chatId: string | number, userId: number): this;
-  public addToScope(type: "chat_administrators", chatId: string | number): this;
-  public addToScope(type: BotCommandScope["type"], chatId?: string | number, userId?: number): this {
-    const scope: BotCommandScope | null = match({ type, chatId, userId })
+  public addToScope(scope: BotCommandGroupsScope, ...middleware: ChatTypeMiddleware<C, "group" | "supergroup">[]): this;
+  public addToScope(scope: BotCommandScopeAllPrivateChats, ...middleware: ChatTypeMiddleware<C, "private">[]): this;
+  public addToScope(scope: BotCommandScope, ...middleware: Array<Middleware<C>>): this;
+  public addToScope(scope: BotCommandScope, ...middleware: Array<Middleware<C>>): this {
+    match(scope)
```

Comment:
It seems to me like this is the only place where you rely on this dependency. You can completely get rid of it by using a switch statement over the type variable here, without losing much readability. In fact, your code might be easier to understand because people don't have to first guess how this library works.

Reply:
I can't specify `{ type: "chat_member", chat_id: P.not(P.nullish).select("chatId"), user_id: P.not(P.nullish).select("userId") }` on a switch statement tho. This is the greatest advantage of using this library for me.
I could move to something like a strategy pattern, but I like the way this library works, and I'd honestly rather keep it then refactor right now.

Reply:
```
case "chat_member:
  if (chat_id && user_id)
```
but okay, fine by me to keep the dep